# Week 4 reading report
Read README.md. Keep units in your explanation.


<!-- assignment-workflow -->
## Before you begin
Open ASSIGNMENT_INSTRUCTIONS.txt from your extracted folder for the starting files, steps, checks and exact upload list. Read STUDENT_HOW_TO.md in a text editor for account creation and Colab/download help. Keep the required names below unchanged. Write your own answers; guided_practice.ipynb is not a graded upload.


In [1]:
STUDENT_ID="24/1035/BSCE-S"
COHORT="CML3112"
import csv
with open("site_inspection_log.csv",newline="") as f:
    rows=list(csv.DictReader(f))


## valid_count (5 marks)
Store an integer count of rows whose crack_width_mm field is nonblank. Skip blank readings before numeric conversion.


In [2]:
valid_count = 0
for row in rows:
    if row['crack_width_mm'] != '':
        valid_count += 1
print(valid_count)


297


## Interesting Pattern
  This code counts how many crack width readings are not blank. It starts a counter at 0, goes through every row, and adds 1 whenever the reading is not empty. Blank readings are skipped, so the result is 297 out of 300 rows.

## mean_reading (5 marks)
Store the arithmetic mean of the nonblank crack_width_mm values as a number. Add the valid readings and divide by their count. Keep full precision.


In [3]:
total_width = 0
for row in rows:
  if row['crack_width_mm'] != '':
    total_width += float(row['crack_width_mm'])
mean_width = total_width / valid_count
print(mean_width)


0.2262996632996631


## Interesting Pattern
This code finds the average crack width. It adds up all the nonblank readings, converting each to a number with float, then divides the total by valid_count (297). Blanks are left out of both the total and the divisor. The result is 0.2262996632996631, kept at full precision.

## above_limit (5 marks)
Store an integer count of valid crack_width_mm readings strictly greater than 0.3. A reading equal to the threshold is not counted.


In [4]:
above_limit = 0
for row in rows:
  if row['crack_width_mm'] != '' and float(row['crack_width_mm']) > 0.3:
    above_limit += 1
print(above_limit)


59


## Interesting Pattern
This code counts how many valid crack width readings are greater than 0.3 mm. It starts a counter at 0 and goes through every row. For each row, it first checks that the reading is not blank, then converts it to a number and checks if it is strictly greater than 0.3. If both conditions are true, it adds 1 to the counter. Blank readings are skipped, and a reading of exactly 0.3 is not counted. The final result is 59.

## group_means (5 marks)
Store a dictionary whose keys are the structure_id labels and whose values are the mean valid crack_width_mm readings within each group. Divide each group total by its own valid count. Keep full precision.


In [5]:
group_totals = {}
group_counts = {}

# Accumulate totals and counts for each group, skipping blank values
for row in rows:
  structure_id = row['structure_id']
  crack_width_mm = row['crack_width_mm']
  if crack_width_mm != '':
    val = float(crack_width_mm)
    if structure_id not in group_totals:
      group_totals[structure_id] = 0.0
      group_counts[structure_id] = 0
      group_totals[structure_id] += val
      group_counts[structure_id] += 1

      # compute the mean for each group with full precision
      group_means ={}
      for struct_id in group_totals:
        group_means[struct_id] = round(group_totals[struct_id] / group_counts[struct_id], 2)

print(group_means)


{'B1': 0.09, 'B2': 0.15, 'B3': 0.19, 'B4': 0.15, 'B5': 0.17, 'B6': 0.2, 'B7': 0.24, 'B8': 0.3, 'B9': 0.31, 'B10': 0.28, 'B11': 0.3, 'B12': 0.32}


## Interesting Pattern
This code finds the average crack width for each structure (B1, B2, B3...). It keeps two dictionaries: one with the running total of readings per structure, and one with how many valid readings each structure has. For every row, it skips blank readings, creates a new entry if the structure is new, then adds the reading to the total and 1 to the count. At the end, it divides each structure's total by its own count to get its mean.

Remove round(..., 2), since the question says to keep full precision.